### PR TITLE
Add support for range aggregations on histogram mapped fields

### LINKED ISSUE #74071: Add range aggregation support for histogram mapped fields
When querying against a histogram field the following occurs:

With a range aggregation:

```
"Field [transaction.duration.histogram] of type [histogram] is not supported for aggregation [range]"

```

With a filters aggregation containing a range query:
```
"Field [transaction.duration.histogram] of type [histogram] does not support range queries"
```

It would be beneficial for `histogram` mapped fields to support range aggregations and queries.

In my mind, it seems that a query matching against the `histogram.values` and then using the corresponding `histogram.counts` entry seems like it should work. 

Issues that pop into my mind are:

- Ranges that span multiple histogram buckets
- Ranges that span a partial bucket
- Multiple ranges that are contained within the same value bucket

I am not sure if these issues prohibit this from working or not.

### PR DESCRIPTION
**Pull request by benwtrent**
This adds support for the range aggregation over `histogram` mapped fields.

Decisions made for implementation:

 - Sub-aggregations are not allowed. This is to simplify implementation and follows the prior art set by the `histogram` aggregation
 - Nothing fancy is done with the ranges. No filter translations as we cannot easily do a `range` filter query against histogram fields. This may be an optimization in the future.
 - Ranges check the histogram value ONLY. No interpolation of values is done. If we have better statistics around the histogram this MAY be possible.

closes https://github.com/elastic/elasticsearch/issues/74071

### CODE REVIEW

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/bucket/range/HistoBackedRangeAggregator.java [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+        InternalRange.Factory<?, ?> rangeFactory,
+        RangeAggregator.Range[] ranges,
+        boolean keyed,
+        AggregationContext context,
+        Aggregator parent,
+        CardinalityUpperBound cardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        final double avgRange = ((double) context.searcher().getIndexReader().maxDoc()) / ranges.length;
+        return hasOverlap(ranges) ? new Overlap(
```
Comment (not-napoleon) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/bucket/range/HistoBackedRangeAggregator.java:45]: I'm nitpicking, but 25 lines is too long for the `?:` operator, IMHO.  Please make this an `if {} else {}` block.

Comment (benwtrent) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/bucket/range/HistoBackedRangeAggregator.java:45]: hahaha! No worries :)

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/bucket/histogram/HistoBackedHistogramAggregator.java [hunk 2]
```diff
@@ -86,9 +86,9 @@ public void collect(int doc, long owningBucketOrd) throws IOException {
                             } else {
                                 collectBucket(sub, doc, bucketOrd);
                             }
-                            // We have added the document already. We should increment doc_count by count - 1
+                            // We have added the document already. We should increment doc_count by count - docCountProvider
                             // so that we have added it count times.
-                            incrementBucketDocCount(bucketOrd, count - 1);
+                            incrementBucketDocCount(bucketOrd, count - docCountProvider.getDocCount(doc));
```
Comment (csoulios) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/aggregations/bucket/histogram/HistoBackedHistogramAggregator.java:91]: Looks like we both fixed the same bug. I merged #74650, so will have a conflict here.

**Review by csoulios — CHANGES_REQUESTED** [review]
Generally, this is a great job done. I loved the accuracy tests :)

I have left couple small comments.

**Review by csoulios — APPROVED** [review]
LGTM!  Thank you for contributing this

**Pull request merged**